init apache hadoop package

kranurag7 opened the pull request:
Fixes:

Related:

### Pre-review Checklist

<!--
This checklist is mostly useful as a reminder of small things that can easily be
forgotten – it is meant as a helpful tool rather than hoops to jump through.

At the moment of this PR you have the most information on what all the change
will affect, so please take the time to jot it down.

Put an `x` in all the items that apply, make notes next to any that haven't been
addressed, and remove any items that are not relevant to this PR.

-->

#### For new package PRs only
<!-- remove if unrelated -->
- [ ] This PR is marked as fixing a pre-existing package request bug
  - [ ] Alternatively, the PR is marked as related to a pre-existing package request bug, such as a dependency
- [x] REQUIRED - The package is available under an OSI-approved or FSF-approved license
- [ ] REQUIRED - The version of the package is still receiving security updates
- [ ] This PR links to the upstream project's support policy (e.g. `endoflife.date`)

kranurag7 committed:
init apache hadoop package

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
CVEs fix

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
pass in properties via properties-file

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
bump deps

this commit bump the deps and also comments out the ones that are not
compiling.

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
add nodejs and yarn

for some reason build on x86 is working good but not for arch so adding
these two as build dependencies. This is required because we are
building yarn-ui.

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
build only for x86 arch

frontend-maven-plugin which is required for building the UI doesn't work
on aarch64 architecture so we are only building for x86 arch.

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
some more CVE fixes

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

EyeCantCU commented on hadoop.yaml:
```diff
@@ -0,0 +1,86 @@
+    - x86_64
+  dependencies:
+    runtime:
+      - openjdk-8
```
We don't need the full JDK. Just `openjdk-8-default-jvm`

EyeCantCU commented on hadoop.yaml:
```diff
@@ -0,0 +1,86 @@
+      - maven
+      - nodejs
+      - openjdk-8
+      - wolfi-base
+      - wolfi-baselayout
```
We don't need these

EyeCantCU commented on hadoop.yaml:
```diff
@@ -0,0 +1,86 @@
+environment:
+  contents:
+    packages:
+      - bash
```
We can remove this

EyeCantCU commented on hadoop.yaml:
```diff
@@ -0,0 +1,86 @@
+
+  - runs: |
+      tar --version
+      mkdir -p ${{targets.contextdir}}/
```
The destination directory (destdir) already exists. We only need to do this for subpackages. Not to mention, this is handled by the command below it

EyeCantCU commented on hadoop.yaml:
```diff
@@ -0,0 +1,86 @@
+      mvn -Pdist,src,yarn-ui -DskipTests -Dtar -Dmaven.javadoc.skip=true clean package -e -X
+      tar -xvf hadoop-dist/target/hadoop-${{package.version}}.tar.gz -C ${{targets.contextdir}}/usr/share/java
+      for f in ${{targets.contextdir}}/usr/share/java/hadoop-${{package.version}}/bin/*; do
+        ln -sf $f ${{targets.contextdir}}/usr/bin/$(basename $f)
```
This is no-op. It's going to create links to files that don't exist as contextdir will be accounted for

EyeCantCU commented on hadoop.yaml:
```diff
@@ -0,0 +1,86 @@
+  pipeline:
+    - name: version check
+      runs: |
+        /usr/share/java/hadoop-${{package.version}}/bin/hadoop version
```
This should be `/usr/share/java/hadoop`, without the package version

kranurag7 replied:
the following line extracts the built hadoop tarball at `/usr/share/java` and the main hadoop binary lives at `/usr/share/java/hadoop-3.4.0/bin/hadoop` that's being used in tests. Please let me know if I should change something here in terms of placing the binary at some other location either directly or via symlink. 

```bash
tar -xvf hadoop-dist/target/hadoop-${{package.version}}.tar.gz -C ${{targets.destdir}}/usr/share/java
```

EyeCantCU replied:
We should definitely exclude version in the path. I don't see a reason to have multiple versioned installations of Hadoop and this will complicate things when we set the working directory as well need to pass the Hadoop version

EyeCantCU commented on hadoop/pombump-properties.yaml:
```diff
@@ -0,0 +1,20 @@
+properties:
+  # the following deps can't be bumped because of signature errors.
+  # - property: avro.version
+  #   value: "1.11.3"
+  # - property: dnsjava.version
+  #   value: "3.6.0":
+  # - property: commons-compress.version
+  #   value: "1.26.0"
```
We can remove these, we'll tackle these when we file advisories

EyeCantCU commented on hadoop.yaml:
```diff
@@ -0,0 +1,86 @@
+    # note: skipping aarch64 because frontend-maven-plugin doesn't work on aarach arch.
+    - x86_64
+  dependencies:
+    runtime:
```
There's bundled bash scripts so we should add `bash` at runtime

kranurag7 committed:
address review comments

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
bash is required to build this package

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
remove versioned directory

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
add hadoop-runner subpackage

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
add more runtime deps

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
replace nmap with netcat-openbsd and add coreutils

we need coreutils because we are missing env binary on latest image.

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
add yarn as runtime deps to hadoop

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
remove yarn again and fix symlink

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
add logging in destdir

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
add curl as runtime dependency to image

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
symlink profiler as well

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

EyeCantCU commented on hadoop.yaml:
```diff
@@ -0,0 +1,115 @@
+    pipeline:
+      - runs: |
+          mkdir -p ${{targets.contextdir}}/opt
+          ln -sf /usr/share/java/hadoop ${{targets.contextdir}}/opt
```
Shouldn't this be `/opt/hadoop`?

kranurag7 committed:
update symlink

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
compile native libraries for performance reasons

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
add more build time deps to compile natively (still failing)

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

EyeCantCU committed:
fix(hadoop): Fix build with OpenSSL > 1.1.1

These are undefined when using OpenSSL > 1.1.1 so the build fails.
Remove the version resitriction so that the common client builds

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(hadoop): Build with snappy

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(hadoop): Build with FUSE support for HDFS

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

kranurag7 committed:
use gnutar and yam hadoop.yaml

using gnutar because busybox tar is missing some flags.

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
add abseil-cpp-dev libraries for native build as well

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
add patch to fix the build and more deps

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
add find and awk as runtime dependencies

some example jobs require this so adding this as runtime dependency.

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
some more build time libraries

these are some additional libraries being added which are optional at
build time but hadoop checknative -a command flags mentions it so adding
it here.

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 commented on hadoop.yaml:
```diff
@@ -0,0 +1,115 @@
+        - jq
+        - async-profiler
+    pipeline:
+      - uses: git-checkout
+        with:
+          repository: https://github.com/apache/hadoop
+          # this branch is where the hadoop-runner base image comes from.
+          # for update please look into the PR that are updating the base image.
+          # previous work: https://github.com/apache/hadoop/pull/6910
+          branch: docker-hadoop-runner-jdk11-u2204
+          expected-commit: 5bb6f450340fa3deb931571d08d42a3f358af8bf
```
given this is a branch where upstream supports docker related configuration. I didn't find a way to poll based on branches. For now, I think this is stable because they are making new changes on another branch but if there's a way then please let me know about it. I'll implement the same. The central point here is that we need to poll on branches that start with `docker-` and then checkout that and build it.
From my attempt of looking into it, I find that doing it manually is the way to go for now because they also remove something in new branches which we depend on. 
For example: This (ref: https://github.com/apache/hadoop/tree/docker-hadoop-3.4) is latest `docker-` branch but python configuration on which we depend is missing from this branch.

EyeCantCU replied:
I'm not sure I follow. We'd create a separate package and then just poll like this:

```yaml
update:
  enabled: true
  schedule: daily
  git: {}
```

We shouldn't need to pass anything else as it will leverage whatever we pass to `git-checkout` for polling

kranurag7 replied:
thank you for the review. New package is tracked [here](https://github.com/wolfi-dev/os/pull/30294) and I've included the same as runtieme dependency to this package (https://github.com/wolfi-dev/os/pull/30294)

EyeCantCU commented on hadoop.yaml:
Rename as `hadoop-3.4`

EyeCantCU commented on hadoop.yaml:
```diff
@@ -0,0 +1,147 @@
+package:
+  name: hadoop
```
Update the name here too

EyeCantCU commented on hadoop.yaml:
```diff
@@ -0,0 +1,147 @@
+      - protoc~3.21
+      - snappy-dev
+      - snappy-static
+      - zlib
```
`zlib-dev` pulls in `zlib`

EyeCantCU commented on hadoop.yaml:
```diff
@@ -0,0 +1,147 @@
+
+  - runs: |
+      mkdir -p ${{targets.destdir}}/usr/share/java/hadoop
+      mkdir -p ${{targets.destdir}}/usr/bin
```
This folder is empty, we can remove this

EyeCantCU commented on hadoop/fix-build-openssl.patch:
Drop this

kranurag7 committed:
hadoop-3.4 and add hadoop-runner as runtime deps

CI will fail.

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
use hadoop-runner as runtime

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 commented:
CI is failing because `hadoop-runner` is not there which was include as runtime dependency to this package. 

Logs: 

`failed to bundle: unable to build graph: hadoop-3.4-3.4.0-r0: unable to resolve dependency hadoop-runner-3.4: could not find package that provides hadoop-runner-3.4 in indexes`

kranurag7 committed:
build for aarch64 arch

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
skip target-architecture in configuration

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 commented:
closing given this is not longer tracked here.
